adr/decisions/2025-06-26-delete-keys.md

```diff
@@ -0,0 +1,93 @@
+* 🟩 **Good**, Using unsafe terminology denotes that this is potentially a harmful operation
+* 🟩 **Good**, Most simple solution
+* 🟥 **Bad**, No way of recovering your private key if not backed up before deletion.
+
+### Option 2: Environment Variable Gate
+
+We can introduce an environment variable, such as OPENTDF_ENABLE_KEY_DELETION, that acts as a gate. When this variable is set to true and the OpenTDF Platform is restarted, the key deletion functionality would become accessible. This approach ensures that key deletion is not enabled by default and requires a deliberate, conscious action from an administrator.
+
+* 🟩 **Good**, Explicit Opt-In: Demands a clear and intentional administrative decision to enable a high-risk operation.
+* 🟩 **Good**, Reduced Accidental Deletion: The required platform restart provides a natural "cooling-off" period, reducing the likelihood of impulsive or accidental deletions.
+* 🟩 **Good**, Simple Implementation: Relatively straightforward to implement in our existing configuration management.
+* 🟨 **Neutral**, Operational Friction: The necessity of a platform restart might be inconvenient for some customers, particularly in high-availability environments.
+* 🟨 **Neutral**, Another environment variable to be added to the platform.
+* 🟥 **Bad**, No way of recovering your private key if not backed up before deletion.
+
+### Option 3: HashiCorp-Inspired Soft Delete (Versioned KV Store)
+
+Drawing inspiration from HashiCorp's KV secrets engine, we could implement a "soft delete" mechanism if our key storage backend supports versioning. Instead of physically removing the key data, the delete command would mark the key as deleted and prevent it from being returned in standard get requests. The historical record of the key, including its deletion status, would be retained.
+
+* 🟩 **Good**, Enhanced Auditability: Retains a historical record of all keys, even those "deleted," which can be invaluable for forensic analysis and compliance.
+* 🟩 **Good**, Reversible (in theory): While the data remains inaccessible, the underlying key might technically be "undeleted" in an emergency (though this would be a complex and highly risky operation itself).
+* 🟨 **Neutral**, Implementation Complexity: Requires a robust, versioned key-value storage backend, which might necessitate significant architectural changes if not already in place.
+* 🟨 **Neutral**, Misleading "Deletion": While marked as deleted, the key data still exists, which might not align with an administrator's expectation of true "deletion" for sensitive information.
+* 🟨 **Neutral**, Storage Overhead: Retaining deleted key versions increases storage requirements.
+* 🟥 **Bad**, Unclear behavior when a key access server is to be deleted and all keys are marked as *deleted*. Should we allow the keys to be deleted? Should we block the deletion of the key access server?
+
+>[!NOTE]
+>A derivative of this option could be that we move the key to an archived table
```

Comment:
Here is another alternative:

1. Allow an optional public key to be configured.  It would be expected the customer has provided the private key to another entity that they trust, or at least a separation of concerns internally.
2. If no key is configured delete is a actual removal
3. If key is configured, instead the key is encrypted by the configured public key

This would fully allow the customer to control the behavior.  Requiring an additional action to enable key retention, but making it clear, and their responsibility to maintain the private key for recovery separately.

Reply:
With the currently encrypted private key, I assume it can be decrypted, meaning the master private key is also present on the system?

The key difference in my solution is that only a public key is configured, so that the encryption key can't be recovered by the key server.